### DISCUSSION TITLE
Publishing issue in VS Marketplace

### DISCUSSION THREAD

**Discussion by zeluspudding** [microsoft/vscode-discussions#832] [Extension Development QnA] [answered]
In short, I'm trying to formally distribute my extension. It works fine when I run it locally but when I try to use the published version from the marketplace, it fails with the following error:

```
workbench.desktop.main.js:sourcemap:2279 Activating extension 'Scious.Bask' failed: Cannot find module 'puppeteer-core'
Require stack:
- c:\Users\Artemis\.vscode\extensions\scious.bask-0.6.2\src\launch.js
- c:\Users\Artemis\.vscode\extensions\scious.bask-0.6.2\src\extension.js
- c:\Program Files\Microsoft VS Code\resources\app\out\vs\loader.js
- c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-amd.js
- c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-fork.js.
```

The node module `puppeteer-core` is listed in my extension's package.json, so I assumed it would just work... but it sadly does not :( I've also tried solutions suggested [here](https://stackoverflow.com/questions/55771578/cant-find-the-module-after-extension-is-installed), [here](https://stackoverflow.com/questions/67191722/vscode-extension-cannot-find-module-when-installed-from-marketplace) and [here](https://stackoverflow.com/questions/34291655/node-modules-not-publishing-for-visual-studio-code-extension) to no avail.

I've recorded a 5 minute video of the issue that [you can watch here](https://www.loom.com/share/60bda143742f4d96abc0e8c949d89c7c?sid=e3d8c82e-d3c0-4cc9-8e18-1fb92763d984) (increase the play back speed to get through it more quickly if you like). Pretty Please help! Thank you in advance,

**Comment by alefragnani** [comment]
Hi @zeluspudding ,

I would suggest you to _package_ the extension (using `vsce package` command) instead of publishing, to better identify whas is missing. Doing so, it will create the `.vsix` file that you can use to `Install from VSIX`. Also, you could use `vsce ls` to _list_ the files that are being packaged. And, if you use Webpack or similar, double check the files.

BTW, I have seen similar issues that were caused because some _dependencies_ should be declared as `devDependency`. I'm not sure this lib is one of these, but worth to look.

Hope this helps

**Comment by zeluspudding** [comment]
Hey Alessandro, thanks for your post! I ran `vsce ls` and found a few files to prune from my bundle - much appreciated! That said, I'm left with the following files:
```
CHANGELOG.md
dist/extension.js
LICENSE.md
package.json
README.md
src/bask_pull.js
src/bask_push.js
src/database.js
src/example_build.js
src/extension.js
src/git.js
src/globals.js
src/img/bask_logo_128x128.png
src/img/Bask.woff
src/img/spa.svg
src/launch.js
src/plugins.js
```
That looks like everything I need - I purposefully exclude the `node module` folder because A) I assume that's created / downloaded at runtime and B) the [official publishing docs](https://code.visualstudio.com/api/working-with-extensions/bundling-extension#publishing) recommend to remove it (makes sense).

So next I followed your recommendation to package to a `.vsix`. Sadly, after installing it, I got the exact same error as before:

```
workbench.desktop.main.js:sourcemap:2279 Activating extension 'Scious.Bask' failed: Cannot find module 'puppeteer-core'
Require stack:
- c:\Users\Artemis\.vscode\extensions\scious.bask-0.6.4\src\launch.js
- c:\Users\Artemis\.vscode\extensions\scious.bask-0.6.4\src\extension.js
- c:\Program Files\Microsoft VS Code\resources\app\out\vs\loader.js
- c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-amd.js
- c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-fork.js.
```

So it still looks like vscode simply isn't installing dependencies... Any other ideas what may be at play? If it helps, below is a copy of my `package.json` - can you confirm that I don't need to include some setup script like "npm install"?

```
{
  "name": "Bask",
  "displayName": "Bask",
  "description": "Frictionless Bubble plugin development. Auto-sync local code changes with Bubble.",
  "icon": "src/img/bask_logo_128x128.png",
  "version": "0.6.4",
  "pricing": "Trial",
  "publisher": "Scious",
  "homepage": "https://docs.scious.io/bask/latest",
  "engines": {
    "vscode": "^1.70.3",
    "node": ">=16"
  },
  "categories": [
    "Other"
  ],
  "activationEvents": [
    "onCommand:bask.pull",
    "onCommand:bask.auto_push",
    "onCommand:bask.auto_build_and_push",
    "onCommand:bask.switch_plugin",
    "onCommand:bask.pause_push"
  ],
  "main": "./src/extension.js",
  "contributes": {
    "commands": [
      {
        "command": "bask.pull",
        "title": "Bask: Pull"
      },
      {
        "command": "bask.auto_push",
        "title": "Bask: Push (auto)"
      },
      {
        "command": "bask.pause_push",
        "title": "Bask: Unlock"
      },
      {
        "command": "bask.switch_plugin",
        "title": "Bask: Switch Plugin"
      }
    ],
    "icons": {
      "bask-lotus": {
        "description": "Bask icon",
        "default": {
          "fontPath": "./src/img/bask.woff",
          "fontCharacter": "\\e900"
        }
      }
    },
    "configuration": {
      "title": "Bask",
      "properties": {
        "bask.sourceFolderName": {
          "type": "string",
          "default": "source",
          "markdownDescription": "The name of the folder Bask will store a plugin's source files in."
        }
      }
    }
  },
  "scripts": {
    "lint": "eslint .",
    "test": "node ./test/runTest.js",
    "vscode:prepublish": "npm run esbuild-base -- --minify",
    "esbuild-base": "esbuild src/extension.js --bundle --outfile=dist/extension.js --external:vscode --format=cjs --platform=node",
    "commit-and-push": "git add . && git commit -m \"Auto commit after bundling\"",
    "esbuild": "npm run esbuild-base -- --sourcemap",
    "esbuild-watch": "npm run esbuild-base -- --sourcemap --watch",
    "test-compile": "tsc -p ./"
  },
  "dependencies": {
    "node-sqlite3-wasm": "0.6.9",
    "puppeteer-core": "19.7.2",
    "simple-git": "3.19.1"
  },
  "devDependencies": {
    "@types/glob": "^8.1.0",
    "@types/mocha": "^10.0.1",
    "@types/node": "20.2.5",
    "@types/vscode": "^1.70.3",
    "@vscode/test-electron": "^2.3.4",
    "chai": "^4.3.7",
    "dotenv": "16.0.3",
    "eslint": "^8.41.0",
    "glob": "^8.1.0",
    "mocha": "^10.2.0",
    "typescript": "^5.1.3"
  }
}
```


Thank you in advance!

**Comment by alefragnani** [comment]
Hi @zeluspudding ,

First of all, thanks for the detailed info.

About the publishing docs, it indeed says to remove the `node_modules` folder, but at the same time, it is considering that _everything is in `dist\extension.js` bundled file_. Maybe your esbuild config is not bundling `puppeteer_core`. 

AFAIK VS Code doesn´t install dependencies when the extension is installed. Every _additional lib_ that your extension requires to work, needs to be packaged with it. 

Hope this helps

**Comment by zeluspudding** [comment]
One thing that I see other projects do that I don't do is to point `main` in `package.json` to the `dist` version of the extension. Like:
[SQLite Viewer](https://github.com/qwtel/sqlite-viewer-vscode/blob/master/package.json) does:
```
"main": "./out/extension.js",
```

and [Github Pull Request](https://github.com/microsoft/vscode-pull-request-github/blob/main/package.json) does:
```
"main": "./dist/extension",
```

Except when I try that in test mode (pressing F5) I get error:
```
workbench.desktop.main.js:sourcemap:810 Error: ENOENT: no such file or directory, open 'c:\Users\Artemis\Documents\Businesses\Scious\projects\bask\dist\node-sqlite3-wasm.wasm'
```

As for my esbuild config, I do believe all of the config is listed right there in the `esbuild-base` command, no?

```
"esbuild-base": "esbuild src/extension.js --bundle --outfile=dist/extension.js --external:vscode --format=cjs --platform=node",
```

I don't have any project level `esbuild.config.js` and was unable to find any equivalent config for my global install of `esbuild` that could be mucking up the build.. very puzzling. I hear you on how everything should get bundled into `/dist/extension.js` - that makes sense... just struggling to see how it's not :/

**Comment by zeluspudding** [comment, marked as answer]
Got the fix: in `package.json` I had:
```
"main": "./src/extension.js",
```

When I should have been using 
```
"main": "./dist/extension.js",
```

I guess I missed a note somewhere 3 months ago when I started my extension that I should be constantly building to and running from `dist` using the `esbuild-watch` script, lol. Thank you for your help @alefragnani! Your comment

>_it is considering that everything is in dist\extension.js_

got me on the right path. Thanks again!
